### ISSUE TITLE
Apr 2021 - Jun 2021 Notes (Image Acquisition, PDV, CDS, deinterleave)

### ISSUE THREAD

**Issue by john-judge** [john-judge/PhotoZ_upgrades#18]
Notes from my (likely last) meeting with Sarwagya 4/17

Sarwagya left a separate folder on the rig computer with NI-DAQmx enhancements. NI-DAQmx overall approach: create tasks to send signals, trigger them later, in parallel, AI/DO voltage channels in task

6/15 - started working on this list.

TO DO:

- [x] Delete old code / comments, e.g. most of `fillPDout`
- [x] LED frequency -- Sarwagya mentioned that the signal to turn on the LED can possibly cause flickering artifacts. Depending on the camera sampling frequency, I think this could manifest as low-frequency changes in intensity that may be easy to overlook. I will just stay aware of this concern moving forward.  We may be able to confirm in the image readout that intensity / amount of light is reasonable
- [x] Address NI issue (on `takeRLI` call): https://github.com/john-judge/PhotoZ_upgrades/issues/3#issuecomment-862550507
- [ ] ~~Efficiency -- avoid needing to copy images in `Camera::acquireImages`. This means no longer pre-allocating a contiguous block of memory for images, but instead save the pointers that PDV produces when `pdv_wait_image` returns.~~ Wouldn't be a correct approach, see [PDV docs](https://edt.com/api/group__acquisition.html#g121ec47ef3b21399bc429d1b8d3cb114) -- namely multibuf function
- [ ] ~~Rename the DAP variables, functions, and other names to DAQmx~~ Not worth doing
- [x] `acquiOneRecord` -- implement correct behavior of `trialCount` (if `trialCount` is 0, use `pseudoOutput`, typically pre-stimulation data. Based on OG PhotoZ, the original behavior enabled `numSkippedTrials` and `pseudo recording` -- no longer needed, so disabled.
- [x] This last point is more related to `Camera.cpp`, but Sarwagya's opinion is that the parallelized camera readout + deinterleave logic should all be encapsulated within the `Camera` class. 
- [ ] ~~Added Dave class with turbo SM code. Based on Chun's notes and the observed slowness/lack of sync, maybe 'superframing' (chunking 10 or 100 frames per readout) is necessary? In any case, maybe we've run into too many independent issues to justify trying to rewrite Chun's code~~
- [ ] Connect image readout from `MainController::acqui` to GUI -- display the image data.  `DataArray` needs to match the dimensions of the `Camera` output. Note: this was temporarily manipulated in `ArrayWindow.h`:
```
#define DEFAULT_ARRAY_WIDTH 2048//256
#define DEFAULT_ARRAY_HEIGHT 40 //40
```
- [ ] Including moving around quadrants.
- [ ] NI-DAQmx errors from calls in `DapController::acqui` were causing the slow image acquisition apparently
- [ ] capture FP analog input into `buf` in `DapController.cpp`
- [ ] Deinterleave testing and debugging
- [x] Reorder deinterleave and CDS subtract -- for efficiency
- [ ] ~~Write unit tests once everything is working.~~

**Comment by john-judge** [comment]
~~Issue: Sarwagya's transition of `outputs` and `pseudoOutputs` from fstream to uint8 array was not complete in the `main` branch. Need to check on the rig computer and hopefully his development is complete and intact there.~~
Fixed `fillPDout`. Keeping branch `sarwagya_final` for now but probably ok to delete.

**Comment by john-judge** [comment]
## 6/21/21 progress
Stepping through with debugger, these are working fine from `record` > `acquiOneRecord`:
- `arrangeData`: stores input to `raw_diode_data[i][trialNo][j]`
- `getBinnedDiodeTrace` loads a data slot (`Data::rawData`)
- `process` > `dataInversing` > `Data::raw2pro` ends up putting in a constant factor in all data, as affected by `CDS_add`

Then I disabled CDS_subtract, deinterleave, and other camera-related processing. The arrays were still blank.
So in other words, the camera readout itself is now blank?

Next step: print out the camera readout directly to file, like previously.

Tried reverting to previous version of PhotoZ Little Dave; promptly BSOD-crashed.

**Comment by john-judge** [comment]
## 6/22/21 progress
Working version! reverted state of `takeRLI`

Data taken to reassemble and compare locally [here](https://drive.google.com/drive/folders/1KiebjhHO_uhym0sPxrntXSTKhynkcBLO?usp=sharing)

![image](https://user-images.githubusercontent.com/40705003/123002061-4c96ac80-d366-11eb-82c7-30051ff1843e.png)

Quadrants remapped to their respective places and also with quadrant-specific min-max normalization:

![Figure_1](https://user-images.githubusercontent.com/40705003/123003811-5c6bbc00-d379-11eb-965f-93450080a9c6.png)

TurboSM Live image for validation. This is a much taller image.
![image](https://user-images.githubusercontent.com/40705003/122992849-726a8400-d35b-11eb-8b15-26a2cce1b2f5.png)

![image](https://user-images.githubusercontent.com/40705003/123004077-c1bfad00-d379-11eb-8fe6-b0fc5575f00e.png)

Next: 

- [x] Attempt with more light for additional validation if possible
- [ ] Display RLI in background in PhotoZ ArrayWindow
- [ ] Change dimension of image by manipulating camera settings

**Comment by john-judge** [comment]
## 6/23/21 Progress

LED goes on at Frame 200
The LED light should probably saturate the field of view

![readout-RLI-full-out0](https://user-images.githubusercontent.com/40705003/123137649-d3589180-d419-11eb-99b3-f6540df44eb5.png)
![readout-RLI-full-out150](https://user-images.githubusercontent.com/40705003/123137666-d6ec1880-d419-11eb-8981-ae11b8f7fdca.png)
![readout-RLI-full-out350](https://user-images.githubusercontent.com/40705003/123137672-d8b5dc00-d419-11eb-8817-ca93297c507b.png)
![readout-RLI-full-out450](https://user-images.githubusercontent.com/40705003/123137675-d8b5dc00-d419-11eb-9337-1e9c9914b6c9.png)

Ironed out a few bugs with data alignment. There's not much more validation I can do at this point. Next: investigate .cfg files and see if possible to manipulate image dimension -- try to discover how this is done in TurboSM.

## TurboSM config file process

in sm.cpp, `readCamConfig()` pulls in the file `sm_config.dat` -- the first setting, which is:
```
{ "2kx2k_NEURO_BINNED_d_",	{ 2048, 2048, 2048, 1024, 1024, 1024, 1024, 1024 },		//width
					{ 512, 200, 50, 160, 32, 15, 128, 40 } },				//height
```
stored in `configLUT`

Then in `SM_initCamera()` for all 2K cameras, this is called:

```
else if (TwoK_flag) {
	if (config_loaded && config_id) {
		for (int i = 0; i < pdv_chan_num; ++i)
			pdv_setsize(pdv_pt[i], configLUT[config_id - 1].width[curConfig], configLUT[config_id - 1].height[curConfig]);
	}
```

## Test: another dimension setting
When I use `pdv_setsize` I can change the PDV dimensions, but then the camera is very slow and reads out a constant (no real image). It must be some setting that is not correct.

Then I removed `pdv_setsize` and used TurboSM to get PDV into the state 2048x512, which is the default for TurboSM. Then I closed TurboSM, opened PhotoZ, and immediately set PhotoZ to the 2048x512 setting. TakeRli ran fast and smooth and produced real data.

It turns out there are a few more things that Chun didn’t explicitly point out:

- It appears that CDS reset “rows” are fixed at 1024px widths, rather than changing width to match that of the image
- It appears the lower quadrants need to be flipped across the horizontal axis – I suppose Chun mentioned this in the quadrant_readout.docx but not in her emails


Next: try it for another setting? What in TurboSM is missing from PhotoZ?

![image](https://user-images.githubusercontent.com/40705003/123159845-57af1280-d422-11eb-8b0f-d2a8b809e37c.png)

![readout-RLI-full-out450](https://user-images.githubusercontent.com/40705003/123163823-c6e33180-d437-11eb-8954-e75f75f8dae8.png)

**Comment by john-judge** [comment]
When quadrant width was 1024px, reset rows were 1024px. Now with quadrant width 2048px, reset rows are still 1024px. Apparently, reset rows seem to be packed into the quadrant width, which might explain why all quadrant widths are multiples of 1024:
```
{ "2kx2k_NEURO_BINNED_d_", { 2048, 2048, 2048, 1024, 1024, 1024, 1024, 1024 },           //width
                           { 512, 200, 50, 160, 32, 15, 128, 40 } },                     //height
```
This also explains some of the seemingly mismatched dimensions. Take for example PhotoZ’s program #7 (which is TurboSM’s program #0):

4 quadrants * (2048 x 512) 

Divide width by 2 for reset data:

1024x512

Multiply both height and width by 2 for quadrants:

2048x1024

Which matches up to the label for this setting: 
200 Hz   2048x1024

Now, this formula doesn’t apply to all of the programs. Which is why I think there must be other hidden factors and settings that change from program to program. This may mean needing to parametrize the deinterleave and subtractCDS code to handle all the programs differently.

You can see why at this point I’m wondering if we’ve only scratched the surface, and that the better alternative is to attempt to use Chun’s code directly (which is what Chun has recommended).

**Comment by john-judge** [comment]
## 6/28/21 Progress
Adjusted PhotoZ internal dimension expectations to match internal settings in .cfg files. The mismatch with the final dimensions is to be expected, I believe -- there may be more processing to do (just a simple crop, at best?)
Disabled deinterleave to test images with CDS subtraction only. Seems to be working reasonably, for all settings (analyzed 48 images total).
Put image-plotting script in repository  https://github.com/john-judge/PhotoZ_Image.git since interactive matplotlib mode is preferred to Colab. 
```
git clone https://github.com/john-judge/PhotoZ_Image.git
```
## 6/29/21 Progress
Repeating same testing as yesterday, but with deinterleave added back in. This will slow processing down.
If there's strange inexplicable behavior, I would remove CDS subtract and test deinterleave alone.
As suspected, it doesn't look reasonable. What's this?

For 1024x40 (final: 5000 Hz  256x60)
![readout-RLI-full-out0](https://user-images.githubusercontent.com/40705003/123869179-48ced100-d8f6-11eb-8b98-f9eb98891be3.png)

Maybe I should try deinterleaving with the height/width of the final image, rather than the internally-configured image?

**Comment by john-judge** [comment]
## 6/30/21 Progress

Developed the idea of the "missing width factor" -- for the 4 highest settings, there's no correction. For the lower settings, there's a divisive factor in just the width of 2 or 4. Perhaps cropping, or maybe another type of correction that affects only width? To search TurboSM.

In any case, there's something up with deinterleave -- none of the settings seem to work with the deinterleave code I developed to work with a single PDV channel readout. Would be great if I could install Anaconda/Jupyter notebook on the Little Dave rig computer -- would help guesswork efficiency.

```
const char* Camera::LABEL[] = {
"200 Hz   2048x1024", // = 2097152 px
"2000 Hz  2048x100",  // = 204800 px
"1000 Hz  1024x320",  // = 327680 px
"2000 Hz  1024x160",  // = 163840 px
"2000 Hz  512x160",   // = 81920 px
"4000 Hz  512x80",    // = 40960 px
"5000 Hz  256x60",    // = 15360 px
"7500 Hz  256x40" };  // = 10240 px

// The final widths to display
const int Camera::DISPLAY_WIDTH[] =  { 2048, 2048, 1024, 1024, 512, 512, 256, 256 };
const int Camera::DISPLAY_HEIGHT[] = { 1024, 100,  320,  160,  160, 80,  60,  40 };

// The internal quadrant widths based on .cfg files
const int Camera::WIDTH[] = { 2048, 2048, 1024, 1024, 1024, 1024, 1024, 1024 };
const int Camera::HEIGHT[] = { 512, 50,  160,   80,  80,   40,   30,  20 };

const int Camera::FREQ[] = { 200, 2000, 1000, 2000, 2000, 4000, 5000, 7500 };
const char* Camera::PROG[] = { //size size factor to display size
"DM2K_2048x512.cfg", // = 1048576 px 1x width 2x height
"DM2K_2048x50.cfg",  // = 102400 px 1x width 2x height
"DM2K_1024x160.cfg", // = 163840 px 1x width 2x height
"DM2K_1024x80.cfg",  // = 81920 px 1x width 2x height
"DM2K_1024x80.cfg",  // = 81920 px 1/2x width 2x height
"DM2K_1024x40.cfg",  // = 40960 px 1/2x width 2x height
"DM2K_1024x30.cfg",  // = 30720 px 1/4x width 2x height
"DM2K_1024x20.cfg"   // = 20480 px 1/4x width 2x height
};

// CDS subtraction => width is divided by a factor of 2
// Quadrants (4 PDV channels) => width and height are multiplied by a factor of 2
// Then all heights match final heights. However, there's a remaining width factor by which to divide:
const int Camera::WIDTH_DIVISIVE_FACTOR[] = { 1, 1, 1, 1, 2, 2, 4, 4 };
```

Need to rearrange CDS rows before deinterleaving!
The top and bottom rows are read out first, then so on so that the two middle frames are read out last. The rows need to be "deinterlaced" after CDS subtraction and before deinterleaving.

Comments while analyzing the pertinent ~30 lines from makeLookupTable in TurboSM:
```
	#pragma omp parallel for private(row,cols,segment,rowIndex,srcIndex,destIndex)
	// Appears that 'segments' are what I call 'quadrants' ?
	// dwid = file_width / stripes, and 'stripes' is half the number of segments
	for (segment = 0; segment < stripes; ++segment) { // going through segments in range [0,stripes]
		//if (num_pdvChan == 4)
		srcIndex += (image_length / 4)*(channelOrder[segment] / 4); // jump to the quadrant start 
		for (row = 0, rowIndex = 0; row < rows; ++row, rowIndex += file_width) {
			destIndex = rowIndex + dwid * segment; // jumps to the correct row for destination?
			for (cols = dwid; cols > 0; --cols, ++destIndex) {
				Lut[destIndex] = srcIndex;
				srcIndex += SEGS / num_pdvChan;
			}
			if (CDS)
				srcIndex += dwid * SEGS / num_pdvChan;
		}
	}
	#pragma omp parallel for private(row,cols,segment,rowIndex,srcIndex,destIndex)
	for (segment = stripes; segment < SEGS; ++segment) { // going through segments in range [stripes,SEGS]
		// if (num_pdvChan == 4)
		srcIndex += (image_length / 4)*(channelOrder[segment] / 4); // jump to the quadrant start 
		// Now we are going down in file width -- back up? due to CDS reset row interlacing?
		for (row = 0, rowIndex = file_length - file_width; row < rows; ++row, rowIndex -= file_width) {
			destIndex = rowIndex + dwid * (segment - SEGS / 2);
			for (cols = dwid; cols > 0; --cols, ++destIndex) {
				Lut[destIndex] = srcIndex;
				srcIndex += SEGS / num_pdvChan;
			}
			if (CDS)
				srcIndex += dwid * SEGS / num_pdvChan;
		}
	}
```

This is TurboSM on the edge of a fluorescent tape arrow:
![image](https://user-images.githubusercontent.com/40705003/124034265-d6222c00-d9af-11eb-873c-fa1dabbdeeec.png)

This is TurboSM when not seeing anything:
![image](https://user-images.githubusercontent.com/40705003/124034405-023dad00-d9b0-11eb-8c97-f3d01ec529a5.png)

Now compare that to PhotoZ (0 and 150 dark, 350 and 450 light):
![readout-RLI-remappedfull-out450](https://user-images.githubusercontent.com/40705003/124034670-5ea0cc80-d9b0-11eb-97c6-549f26a60080.png)
![readout-RLI-remappedfull-out350](https://user-images.githubusercontent.com/40705003/124034671-5ea0cc80-d9b0-11eb-96a0-37928c1e273f.png)
![readout-RLI-remappedfull-out150](https://user-images.githubusercontent.com/40705003/124034673-5ea0cc80-d9b0-11eb-96c5-b434162a85b8.png)
![readout-RLI-remappedfull-out0](https://user-images.githubusercontent.com/40705003/124034674-5f396300-d9b0-11eb-9123-ac87ee1ed43f.png)

**Comment by john-judge** [comment]
## 7/1/21 Progress

Testing with 2048x1024 (it's the most easily interpretable setting), and temporarily making such that 1) Frame 0 has light on and 2) only 1 frame at a time, to improve efficiency of iteration:
- Tried `stripes = 8`, no change
- Reverted `stripes = 4` but kept `channelOrder = twokchannel; num_pdvChan = 4;` still no change
- Realized there's a bug in pointers to images to process; corrected and reverted to original settings (`stripes = 4`). Different result, not easily troubleshootable
- Reverted to my version of deinterleave
- Changing channel order for ipdv=0 to 0,1,2,3
- Disabled CDS-row deinterlace -- it doesn't seem correct based on result

![readout-RLI-remappedfull-out0](https://user-images.githubusercontent.com/40705003/124170215-00312800-da5c-11eb-9000-0048f278068f.png)
channels to 0,1,2,3

- Changing channel order for all PDV -- no effect
- Changing up the width/height fed to `deinterleave` by factors of 2, preserving image size
- implementing horiz/vert quad flipping and assigning to the correct quadrants. Reconsidered PDV order, revised Python `remapQuadrants` function
- Iterating through different `channelOrder`s
- Camera stopped taking data. Restarted, no change. Power off and wait several minutes, then power on. No change. Restart computer. Start TurboSM first -- camera working again.
- Needed to divide width by 2! Since CDS subtract! Now image is looking reasonable!
![readout-RLI-remappedfull-out0](https://user-images.githubusercontent.com/40705003/124184924-1c8a9000-da6f-11eb-9486-531091638e4e.png)
 How did I forget that?
![image](https://user-images.githubusercontent.com/40705003/124183886-b8b39780-da6d-11eb-9444-7ade9bab624f.png)
